File: dotnet/mission-to-mars/Rover.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace mission_to_mars
{
    public class Rover : IModuleMartien
    {
        public bool IsPretPourRecuperation { get; private set; }

        public Position Position { get; protected set; }

        private readonly Direction _direction;

        public Rover(Direction direction, Position position)
        {
            _direction = direction;
            Position = position;
        }

        public void ActiverRecuperation()
        {
            this.IsPretPourRecuperation = true;
        }

        public void Avancer()
        {
            Position = _direction switch
            {
                Direction.NORD => Position with { Y = Position.Y + 1 },
                Direction.SUD => Position with { Y = Position.Y - 1 },
                Direction.EST => Position with { X = Position.Y + 1 },
                _ => Position with { X = Position.Y - 1 },
            };
        }


        public void Reculer()
        {
            Position = _direction switch
            {
                Direction.NORD => Position with { Y = Position.Y - 1 },
                Direction.SUD => Position with { Y = Position.Y + 1 },
                Direction.EST => Position with { X = Position.Y - 1 },
                _ => Position with { X = Position.Y + 1 },
            };
        }

        public virtual void Descendre()
        {
            throw new NotImplementedException("Un rover ne vole pas");
        }

        public virtual void Monter()
        {
            throw new NotImplementedException("Un rover ne vole pas");
        }

    }
}

File: typescript/src/Rover.ts
import { Direction } from './Direction'
import { ModuleMartien } from './ModuleMartien'
import { Position } from './Position'

export class Rover implements ModuleMartien {
  private pretPourRecuperation: boolean = false;
  private readonly _direction: Direction;
  protected _position: Position;

  constructor(readonly direction: Direction, readonly position: Position) {
    this._direction = direction;
    this._position = position;
  }

  public avancer() {
    switch (this._direction) {
      case Direction.NORD:
        this._position = new Position(this._position.x, this._position.y + 1, this._position.z);
        break;

      case Direction.SUD:
        this._position = new Position(this._position.x, this._position.y - 1, this._position.z);
        break;

      case Direction.OUEST:
        this._position = new Position(this._position.x - 1, this._position.y, this._position.z);
        break;

      default:
        this._position = new Position(this._position.x + 1, this._position.y, this._position.z);
    }
  }

  public reculer() {
    switch (this._direction) {
      case Direction.NORD:
        this._position = new Position(this._position.x, this._position.y - 1, this._position.z);
        break;

      case Direction.SUD:
        this._position = new Position(this._position.x, this._position.y + 1, this._position.z);
        break;

      case Direction.OUEST:
        this._position = new Position(this._position.x + 1, this._position.y, this._position.z);
        break;

      default:
        this._position = new Position(this._position.x - 1, this._position.y, this._position.z);
    }
  }

  isPretPourRecuperation(): boolean {
    return this.pretPourRecuperation;
  }

  getPosition(): Position {
    return this._position;
  }

  monter() {
    throw new Error('Un rover ne vole pas');
  }

  descendre() {
    throw new Error('Un rover ne vole pas');
  }

  activerRecuperation() {
    this.pretPourRecuperation = true;
  }
}
